File: 0-inverted-index-and-bm25-from-scratch/1-java/src/main/java/com/starci/bm25/search/SnapshotStore.java
package com.starci.bm25.search;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.Jedis;

/** Persists/restores the whole index as one JSON value under a single Redis key. */
@Component
public class SnapshotStore {

    private static final String SNAPSHOT_KEY = "bm25:index:snapshot";

    private final JedisPool pool;
    private final ObjectMapper mapper = new ObjectMapper();

    public SnapshotStore(@Value("${redis.host}") String host,
                         @Value("${redis.port}") int port) {
        this.pool = new JedisPool(host, port);
    }

    /** Save on every write so the snapshot is always current. */
    public void save(InvertedIndex.Snapshot snap) {
        try (Jedis jedis = pool.getResource()) {
            jedis.set(SNAPSHOT_KEY, mapper.writeValueAsString(snap));
        } catch (Exception ignored) {
            // Redis outage -> fall back to a pure in-memory index.
        }
    }

    /** Load on boot; returns null when the key is absent or Redis is unreachable. */
    public InvertedIndex.Snapshot load() {
        try (Jedis jedis = pool.getResource()) {
            String raw = jedis.get(SNAPSHOT_KEY);
            if (raw == null) {
                return null;
            }
            return mapper.readValue(raw, InvertedIndex.Snapshot.class);
        } catch (Exception ignored) {
            return null;
        }
    }
}

File: 0-inverted-index-and-bm25-from-scratch/2-csharp/SnapshotStore.cs
using System.Text.Json;
using StackExchange.Redis;

namespace Bm25SearchService;

/// <summary>Persists/restores the whole index as one JSON value under a single Redis key.</summary>
public class SnapshotStore
{
    private const string SnapshotKey = "bm25:index:snapshot";
    private readonly IConnectionMultiplexer? _redis;

    public SnapshotStore(string host, int port)
    {
        try
        {
            var options = new ConfigurationOptions
            {
                EndPoints = { { host, port } },
                AbortOnConnectFail = false,
                ConnectRetry = 3
            };
            _redis = ConnectionMultiplexer.Connect(options);
        }
        catch
        {
            _redis = null; // fall back to a pure in-memory index
        }
    }

    public void Save(Snapshot snap)
    {
        try
        {
            _redis?.GetDatabase().StringSet(SnapshotKey, JsonSerializer.Serialize(snap));
        }
        catch
        {
            // Redis outage -> keep serving from memory.
        }
    }

    public Snapshot? Load()
    {
        try
        {
            var raw = _redis?.GetDatabase().StringGet(SnapshotKey);
            if (raw is null || raw.Value.IsNullOrEmpty) return null;
            return JsonSerializer.Deserialize<Snapshot>(raw.Value!);
        }
        catch
        {
            return null;
        }
    }
}
